cluster/gce/gci/configure-helper.sh

```diff
@@ -83,18 +161,121 @@ function safe-format-and-mount() {
   mount -o discard,defaults "${device}" "${mountpoint}"
 }
 
-# Local ssds, if present, are mounted at /mnt/disks/ssdN.
+# Gets a devices UUID and bind mounts the device to mount location in
+# /mnt/disks/by-id/
+function unique-uuid-bind-mount(){
+  local mountpoint=$1
+  local bindpoint=$2 
+
+  # Mount device to the mountpoint
+  mkdir -p "${bindpoint}"
+  echo "Binding '${mountpoint}' at '${bindpoint}'"
+  mount --bind "${mountpoint}" "${bindpoint}"
+  chmod a+w ${bindpoint}
+}
+
+
+# Mounts, bindmounts, or symlinks depending on the interface and format
+# of the incoming device
+function mount-ext(){
+  local ssd=$1
+  local interface=$2
+  local format=$3  
+
+  # TODO: Handle partitioned disks. Right now this code just ignores partitions
+  if [[ ${format} == "fs" ]]; then
+    if [[ ${interface} == "scsi" ]]; then
+      local scsinum=`echo ${ssd} | sed -e 's/\/dev\/disk\/by-id\/google-local-ssd-\([0-9]*\)/\1/'`
+      local actual_device=$(readlink -f ${ssd} | cut -d '/' -f 3)
+      if [[ -z ${scsinum} ]]; then
+        echo "Failed to get the local disk number for device ${ssd}" >&2
+        exit 2
+      fi
+      local mountpoint="/mnt/disks/ssd${scsinum}"
+    else
+      # This path is required because the existing Google images do not
+      # expose NVMe devices in /dev/disk/by-id so we are using the /dev/nvme instead
+      local nvmenum=`echo ${ssd} | sed -e 's/\/dev\/nvme0n\([0-9]*\)/\1/'`
+      local actual_device=$(echo ${ssd} | cut -d '/' -f 3)
+      if [[ -z ${nvmenum} ]]; then
+        echo "Failed to get the local disk number for device ${ssd}" >&2
+        exit 2
+      fi
+      local mountpoint="/mnt/disks/ssd-nvme${nvmenum}"
+    fi
+
+    # actual_device should be the kernel device name e.g. "sdb", "nvme0n1"
+    if [[ ${actual_device} != sd* ]] && [[ ${actual_device} != nvme* ]]; then
+      echo "'actual_device' is not of the correct format. It must be the kernel name of the device, got ${actual_device} instead" >&2
+      exit 1
+    fi
+
+    safe-format-and-mount ${ssd} ${mountpoint}
+
+    # Trigger udev refresh so that newly formatted devices are propagated in by-uuid
+    udevadm control --reload-rules
+    udevadm trigger
+    udevadm settle 
+
+    # grep the exact match of actual device, prevents substring matching
+    local myuuid=$(ls -l /dev/disk/by-uuid/ | grep "/${actual_device}$" | tr -s ' ' | cut -d ' ' -f 9)
+    # myuuid should be the uuid of the device as found in /dev/disk/by-uuid/ 
+    if [[ -z ${myuuid} ]]; then
+      echo "Failed to get a uuid for device ${ssd} when mounting." >&2
+      exit 2
+    fi
+
+    # bindpoint should be the full path of the to-be-bound device
+    local bindpoint="${UUID_MNT_PREFIX}-${interface}-fs/local-ssd-${myuuid}"
+
+    unique-uuid-bind-mount ${mountpoint} ${bindpoint}
```

Comment:
One more issue I realized.  Some users today may be manually unmounting and unformatting the auto-formatted local SSDs to already try using them as raw block devices.  

So we should not blindly create the bindmount otherwise it could break their scripts.  Instead, we should only bindmount if they specify the FS option in NODE_LOCAL_SSDS_EXT.

Reply:
If a user is to request 1 SCSI FS device through NODE_LOCAL_SSDS_EXT and 1 SCSI FS device through NODE_LOCAL_SSDS (old method) theres actually no way to distinguish between the two SSDS as they are the exact same. In this case, the user could theoretically try to unformat and unmount either of the two devices and we have no way of controlling which one is which.

Possible solutions:
1) Don't support concurrent use of both SSD provisioning methods, if they use the new method we bindmount, old method no bindmounts
2) Support concurrent use but distinguish devices gotten from old and new methods somehow, maybe with some kind of naming change (this will break backwards compatibility)

Reply:
I like 1).  It simplifies the implementation and also is easier to understand.

Reply:
Ok, I will make it so that if a user specifies both methods it will throw an error.

Reply:
This piece of code does break when the user has a software RAID array on these disks. We see the following failure after the reboot -- 
```May 29 19:07:25 gke-dev-main-dev-main-ssd-pool-3-46a5190b-bk8l configure-helper.sh[2227]: mount: /mnt/disks/ssd0: unknown filesystem type 'linux_raid_member'.
May 29 19:07:25 gke-dev-main-dev-main-ssd-pool-3-46a5190b-bk8l configure-helper.sh[2227]: Mounting '/dev/disk/by-id/google-local-ssd-0' at '/mnt/disks/ssd0'```